grpc-server/tinkerbell.go

```diff
@@ -2,47 +2,219 @@ package grpcserver
 func (s *server) GetWorkflowContextList(context context.Context, req *pb.WorkflowContextRequest) (*pb.WorkflowContextList, error) {
-	return exec.GetWorkflowContextList(context, req, s.db)
+	wfs, err := getWorkflowsForWorker(s.db, req.WorkerId)
+	if err != nil {
+		return nil, err
+	}
+	wfContexts := []*pb.WorkflowContext{}
+	for _, wf := range wfs {
+		wfContext, err := s.db.GetWorkflowContexts(context, wf)
+		if err != nil {
+			return nil, status.Errorf(codes.Aborted, "Invalid workflow %s found for worker %s", wf, req.WorkerId)
+		}
+		wfContexts = append(wfContexts, wfContext)
+	}
+	return &pb.WorkflowContextList{
+		WorkflowContexts: wfContexts,
+	}, nil
 }
 
 // GetWorkflowActions implements tinkerbell.GetWorkflowActions
 func (s *server) GetWorkflowActions(context context.Context, req *pb.WorkflowActionsRequest) (*pb.WorkflowActionList, error) {
-	return exec.GetWorkflowActions(context, req, s.db)
+	wfID := req.GetWorkflowId()
+	if len(wfID) == 0 {
+		return nil, status.Errorf(codes.InvalidArgument, "workflow_id is invalid")
+	}
+	return getWorkflowActions(context, s.db, wfID)
 }
 
 // ReportActionStatus implements tinkerbell.ReportActionStatus
 func (s *server) ReportActionStatus(context context.Context, req *pb.WorkflowActionStatus) (*pb.Empty, error) {
-	return exec.ReportActionStatus(context, req, s.db)
+	wfID := req.GetWorkflowId()
+	if len(wfID) == 0 {
+		return nil, status.Errorf(codes.InvalidArgument, "workflow_id is invalid")
+	}
+	if len(req.GetTaskName()) == 0 {
+		return nil, status.Errorf(codes.InvalidArgument, "task_name is invalid")
+	}
+	if len(req.GetActionName()) == 0 {
+		return nil, status.Errorf(codes.InvalidArgument, "action_name is invalid")
+	}
+	fmt.Printf("Received action status: %s\n", req)
+	wfContext, err := s.db.GetWorkflowContexts(context, wfID)
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "workflow context not found for workflow %s", wfID)
+	}
+	wfActions, err := s.db.GetWorkflowActions(context, wfID)
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "workflow actions not found for workflow %s", wfID)
+	}
+
+	// We need bunch of checks here considering
+	// Considering concurrency and network latencies & accuracy for proceeding of WF
+	actionIndex := wfContext.GetCurrentActionIndex()
+	if req.GetActionStatus() == pb.ActionState_ACTION_IN_PROGRESS {
+		if wfContext.GetCurrentAction() != "" {
+			actionIndex = actionIndex + 1
+		}
+	}
+	action := wfActions.ActionList[actionIndex]
+	if action.GetTaskName() != req.GetTaskName() {
+		return nil, status.Errorf(codes.FailedPrecondition, "reported task name not matching in actions info")
+	}
+	if action.GetName() != req.GetActionName() {
+		return nil, status.Errorf(codes.FailedPrecondition, "reported action name not matching in actions info")
+	}
+	wfContext.CurrentWorker = action.GetWorkerId()
+	wfContext.CurrentTask = req.GetTaskName()
+	wfContext.CurrentAction = req.GetActionName()
```

Comment:
All these `Get..` have an incorrect way of handling the struct fields. 

I can understand that these are from prior commits, but it can affect your implementation. How do you think we should handle it. ?

Reply:
Marking resolved, based on this [thread](https://github.com/tinkerbell/tink/pull/247#discussion_r468480839).

Reply:
Let me know once you have updated, I will resolve the comments.